File: src/sorts/QuickSort.java
package sorts;

import java.util.Random;

/**
 * Created by Chris Keyser on 3/17/2016.
 * 2-way partition quicksort implementation
 */
public class QuickSort extends Sort {
    public QuickSort(int[] arr) {
        this.arr = arr;
    }

    @Override
    public int[] sort() {
        shuffle();
        quicksort(0, arr.length - 1);
        return arr;
    }

    private void quicksort(int lowerIndex, int higherIndex) {

        int i = lowerIndex;
        int j = higherIndex;
        // take middle number for pivot
        int pivot = arr[lowerIndex + (higherIndex - lowerIndex) / 2];
        // Divide into two arrays
        while (i <= j) {
            //find a number from the left side less than pivot
            while (arr[i] < pivot) {
                i++;
            }

            //find a number from the right side greater than pivot
            while (arr[j] > pivot) {
                j--;
            }

            //swap them
            if (i <= j) {
                swap(i, j);

                i++;
                j--;
            }
        }

        // call quickSort() method recursively
        if (lowerIndex < j)
            quicksort(lowerIndex, j);
        if (i < higherIndex)
            quicksort(i, higherIndex);
    }

    //standard shuffling algorithm
    //used to avoid bad case for qsort of almost-sorted array
    private void shuffle() {
        Random rand = new Random();

        for(int i = 0; i < arr.length; i++) {
            swap(i, rand.nextInt(arr.length));
        }
    }
}

File: ruby/src/sorts/quick_sort.rb
module QuickSort
    class Sorter
        def qsort(arr)
            return arr if arr.length <= 1

            # take random pivot value
            pivot = arr[rand(0...arr.length)]

            i, j = 0, arr.length - 1

            while i < j
                # find a left side element to swap
                while i < arr.count && arr[i] < pivot
                    i += 1
                end

                # find a right side element to swap
                while j >= 0 && arr[j] > pivot
                    j -= 1
                end

                # swap the values if they should be swapped
                if i < j
                    arr[i], arr[j] = arr[j], arr[i]
                end
            end

            left = arr[0, j]
            right = arr[i, arr.length]

            self.qsort(left) + [pivot] + self.qsort(right)
        end
    end

    def QuickSort.sort(arr)
        # copy the array so this is pure
        out = arr.map { |x| x }

        Sorter.new.qsort(out)
    end
end
